File: src/main/java/toker/panel/entity/Faction.java
package toker.panel.entity;

import com.fasterxml.jackson.annotation.JsonBackReference;
import toker.panel.entity.pk.FactionPK;

import javax.persistence.*;
import java.util.Collection;
import java.util.Objects;

@Entity
@Table(name = "faction")
@IdClass(FactionPK.class)
public class Faction {

    private Integer index;
    private Server server;

    private String name;
    private Integer bannerId;

    private Collection<Player> players;

    @Id
    @Column(name = "index", nullable = false)
    public Integer getIndex() {
        return index;
    }

    public void setIndex(Integer index) {
        this.index = index;
    }

    @Id
    @ManyToOne
    @JoinColumn(name = "server_id", referencedColumnName = "id", nullable = false)
    public Server getServer() {
        return server;
    }

    public void setServer(Server server) {
        this.server = server;
    }

    @Column(name = "name", nullable = false, length = 64)
    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    @Column(name = "banner_id", nullable = false)
    public Integer getBannerId() {
        return bannerId;
    }

    public void setBannerId(Integer bannerId) {
        this.bannerId = bannerId;
    }

    @OneToMany(mappedBy = "faction")
    @JsonBackReference
    public Collection<Player> getPlayers() {
        return players;
    }

    public void setPlayers(Collection<Player> players) {
        this.players = players;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Faction faction = (Faction) o;
        return Objects.equals(index, faction.index) &&
                Objects.equals(server, faction.server);
    }

    @Override
    public int hashCode() {
        return Objects.hash(index, server, name, bannerId);
    }

    public interface View {
        interface None {}
    }
}

File: src/main/java/toker/panel/entity/Faction.kt
package toker.panel.entity

import com.fasterxml.jackson.annotation.JsonView
import toker.panel.entity.pk.FactionPK
import javax.persistence.*

@Entity
@Table(name = "faction")
@IdClass(FactionPK::class)
class Faction(
        @Id
        @Column(name = "index", nullable = false)
        var index: Int? = null,

        @Id
        @ManyToOne
        @JoinColumn(name = "server_id", referencedColumnName = "id", nullable = false)
        var server: Server? = null,

        @Column(name = "name", nullable = false, length = 64)
        var name: String? = null,

        @Column(name = "banner_id", nullable = false)
        var bannerId: Int? = null
) {
    @OneToMany(mappedBy = "faction")
    @JsonView(View.Players::class)
    var players: MutableSet<Player>? = null

    interface View {
        interface None
        interface Players
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as Faction
        if (index != other.index) return false
        if (server != other.server) return false
        return true
    }

    override fun hashCode(): Int {
        var result = index ?: 0
        result = 31 * result + (server?.hashCode() ?: 0)
        return result
    }
}